orttraining/orttraining/core/optimizer/transformer_layer_recompute.cc

```diff
@@ -0,0 +1,198 @@
+  const std::unordered_set<std::string> gelu_ops{"Gelu", "BiasGelu", "FastGelu"};
+  const std::unordered_set<std::string> dropout_ops{"Dropout", "BiasDropout"};
+  const std::unordered_set<std::string> layernorm_ops{"LayerNormalization", "SkipLayerNormalization"};
+
+  std::vector<const NodeArg*> layer_start_edges, layer_end_edges;
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+  for (auto node_index : node_topology_list) {
+    auto& node = *graph.GetNode(node_index);
+
+    // Look for start of a transformer layer
+    if ((layernorm_ops.find(node.OpType()) != layernorm_ops.end() ||
+         dropout_ops.find(node.OpType()) != dropout_ops.end()) &&
+        node.GetOutputEdgesCount() == 4) {
+      layer_start_edges.push_back(node.OutputDefs()[0]);
+    }
+
+    // Look for end of a transformer layer
+    if (gelu_ops.find(node.OpType()) != gelu_ops.end()) {
+      auto next_node = node.OutputNodesBegin();
+
+      while (next_node->OutputNodesBegin() != next_node->OutputNodesEnd() &&
+             dropout_ops.find(next_node->OpType()) == dropout_ops.end()) {
+        next_node = next_node->OutputNodesBegin();
+      }
+
+      while (next_node->OutputNodesBegin() != next_node->OutputNodesEnd() &&
+             layernorm_ops.find(next_node->OpType()) == layernorm_ops.end()) {
+        next_node = next_node->OutputNodesBegin();
+      }
+
+      if (layernorm_ops.find(next_node->OpType()) != layernorm_ops.end()) {
+        layer_end_edges.push_back(next_node->OutputDefs()[0]);
+      }
+    }
+  }
+
+  ORT_RETURN_IF_NOT(layer_start_edges.size() == layer_end_edges.size(), "Number of start and end edges doesn't match!");
+
+  start_end_edges.clear();
+  std::cout << "Find " << layer_start_edges.size() << " transformer layers.\n";
+  for (size_t i = 0; i < layer_start_edges.size(); ++i) {
+    start_end_edges.push_back({layer_start_edges[i], layer_end_edges[i]});
+    std::cout << "Start edge: " << layer_start_edges[i]->Name() << " End edge: " << layer_end_edges[i]->Name() << "\n";
+  }
+
+  return Status::OK();
+}
+
+typedef std::set<const Node*, NodeCompare> NodeSet;
+
+std::vector<const Node*> TransformerLayerRecompute::NodesBetweenEdges(Graph& graph, const NodeArg* start, const NodeArg* end) const {
+  // Forward BFS from the start node
+  std::vector<const Node*> start_nodes = graph.GetConsumerNodes(start->Name());
+  NodeSet fw_visited(start_nodes.begin(), start_nodes.end());
+  std::deque<const Node*> fw_queue(start_nodes.begin(), start_nodes.end());
+  while (!fw_queue.empty()) {
+    const Node* n = fw_queue.front();
+    fw_queue.pop_front();
+
+    for (auto node_it = n->OutputNodesBegin(); node_it != n->OutputNodesEnd(); ++node_it) {
+      const Node& node = *node_it;
+      if (fw_visited.find(&node) == fw_visited.end()) {
+        fw_queue.push_back(&node);
+        fw_visited.insert(&node);
+      }
+    }
+  }
+
+  // Reverse BFS from the end node
```

Comment:
>Reverse BFS  [](start = 5, length = 12)

could we have a general BFS function to reuse between forward and reverse?

Reply:
done. 

---
In reply to: [465970282](https://github.com/microsoft/onnxruntime/pull/4526#discussion_r465970282) [](ancestors = 465970282)